### DISCUSSION TITLE
IAR CMake native support

### DISCUSSION THREAD

**Discussion by HiFiPhile** [hathach/tinyusb#2372] [General]
With 9.50.1 release, IAR added experimental native CMake support (strangely not mentioned in public release note). Now it's possible to import `CMakeLists.txt` then build and debug as a normal project.

Following these steps:

1. Add IAR compiler binary path to system `PATH` environment variable, such as `C:\Program Files\IAR Systems\Embedded Workbench 9.2\arm\bin`
2. Get a GNU `size` executable and add the path to system `PATH` environment variable, you can use Mingw for example.
3. Create new project in IAR, in Tool chain dropdown menu, choose `CMake for Arm` then `Import CMakeLists.txt`.
4. Set up board option in `Option - CMake/CMSIS-TOOLBOX - CMake`, for example `-DBOARD=stm32f439nucleo -DTOOLCHAIN=iar`, **Uncheck `Override tools in env`**
5. (For debug only) Choose correct CPU model in `Option - General Options - Target`, to profit register and memory view.

@hathach I got following error and I had to comment out `family_flash_jlink(${TARGET})` in `family.cmake`, do you have any idea ?
```
	-- Check for working CXX compiler: C:/Program Files/IAR Systems/Embedded Workbench 9.2/arm/bin/iccarm.exe - skipped
	-- Detecting CXX compile features
	-- Detecting CXX compile features - done
	-- Configuring done
	CMake Error in CMakeLists.txt:
	  Evaluation file to be written multiple times with different content.  This
	  is generally caused by the content evaluating the configuration type,
	  language, or location of object files:
	
	   D:/mcu/tinyusb/examples/device/audio_4_channel_mic/.cmake_build/audio_4_channel_mic.jlink
```
Full log: [log_cmake.zip](https://github.com/hathach/tinyusb/files/13646418/log_cmake.zip)

**Comment by hathach** [comment]
glad to see IAR is moving to support cmake project. Unfortunately, I don't have IAR workbench (also did't use it), this look like an setup related issue. For family_flash_jlink(), it is an function to create custom target e.g `audio_4_channel_mic-jlink` to flash target using jlink (similar to flash-jlink in makefile). It does generate an audio_4_channel_mic.jlink when invoked, but should only does it once per target. The error suggest that it is called multiple times. I notice there is `Multi-Config` (release/debug) in your cmake invocation, maybe try to drop that first. Normally i have each config build in its own .cmake_build folder.

https://github.com/hathach/tinyusb/blob/59ecdb78fed4221c3e623d7ea9a9e69c3cd87aae/hw/bsp/family_support.cmake#L364-L382
